File: backend/src/main/java/com/example/backend/service/AuthService.java
package com.example.backend.service;

import com.example.backend.model.User;
import com.example.backend.repository.UserRepository;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;

import java.util.Optional;

@Service
public class AuthService {

    private final UserRepository userRepository;
    private final PasswordEncoder passwordEncoder;

    public AuthService(UserRepository userRepository, PasswordEncoder passwordEncoder) {
        this.userRepository = userRepository;
        this.passwordEncoder = passwordEncoder;
    }

    public User register(String name, String email, String rawPassword) {
        if (userRepository.existsByEmail(email)) {
            throw new IllegalArgumentException("Email already in use.");
        }

        User user = new User();
        user.setName(name);
        user.setEmail(email);
        user.setPassword(passwordEncoder.encode(rawPassword));
        return userRepository.save(user);
    }

    public User authenticate(String email, String rawPassword) {
        Optional<User> user = userRepository.findByEmail(email);
        if (user.isEmpty() || !passwordEncoder.matches(rawPassword, user.get().getPassword())) {
            throw new IllegalArgumentException("Invalid credentials.");
        }
        return user.get();
    }
}

File: backend/src/services/authService.js
import bcrypt from "bcryptjs";
import { User } from "../models/user.js";

export async function register(name, email, rawPassword) {
  if (await User.exists({ email })) {
    throw new Error("Email already in use.");
  }

  const password = await bcrypt.hash(rawPassword, 10);
  return User.create({ name, email, password });
}

export async function authenticate(email, rawPassword) {
  const user = await User.findOne({ email });
  if (!user || !(await bcrypt.compare(rawPassword, user.password || ""))) {
    throw new Error("Invalid credentials.");
  }
  return user;
}
